crates/shadowsocks/src/relay/udprelay/compat.rs

```diff
@@ -0,0 +1,88 @@
+use async_trait::async_trait;
+use std::{
+    io::Result,
+    net::SocketAddr,
+    ops::Deref,
+    task::{Context, Poll},
+};
+use tokio::io::ReadBuf;
+
+use crate::net::UdpSocket;
+
+/// a trait for datagram transport that wraps around a tokio `UdpSocket`
+#[async_trait]
+pub trait DatagramTransport: Send + Sync + std::fmt::Debug {
```

Comment:
Are these bounded traits has to be `Send`, `Sync` and `Debug`?

It could be restricted on `ProxySocket<S>`.

Reply:
yes but to use it, one has to impl Sync Send anyways